### PR TITLE
Proxying non-enumerable properties as well

### PR DESCRIPTION
**Pull request by timdp**
We use React with ES6 classes transpiled by Babel. If we try to access a class method on the custom element powered by Reactive Elements, it fails. For example, this doesn't work:

```javascript
// Gets run through Babel

class CompBase extends React.Component {
  foo() {}
}

class MyComp extends CompBase {
  bar() {}
}

document.registerReact('my-comp', MyComp)
```

```html
<my-comp></my-comp>
<script>
document.querySelector('my-comp').foo()
document.querySelector('my-comp').bar()
</script>
```

The underlying reason seems to be that class methods aren't enumerable in ES6, so I wrote this patch that uses `Object.getOwnPropertyNames` to proxy everything along the prototype chain up until `React.Component`. Not sure if it's a good idea, but it works for our scenario.

Any thoughts?

### CODE REVIEW

**Comment by PixelsCommander** [comment]
Have you tried this one with demo in repository? Do all buttons work after?

**Comment by timdp** [comment]
The demo page works fine here.

However, it isn't the greatest test case, because a) it doesn't use ES6/Babel and b) it doesn't use inheritance, so the prototype chain is trivial.

My main concern is copying too many properties. I *think* stopping at `React.Component.prototype` will work in all cases, but it still feels too aggressive somehow.

**Comment by timdp** [comment]
:tada:

**Comment by PixelsCommander** [comment]
=)

**Comment by kof** [comment]
I am sure its not a good idea. The state before was to explicitly expose methods you want to be accessible from the custom element. Merging the whole component will lead  to issues and unintentional side effects.

The way before was easy, we had a props.element. 

So whats so hard with: `this.props.element.myMethod = ::this.myMethod`?

**Comment by kof** [comment]
Accessing imperative methods should be edge case anyways when you are not able to find any better declarative solution.

**Comment by kof** [comment]
You should not have the need to expose ALL methods on the element.

File: dist/reactive-elements.js [hunk 1]
```diff
@@ -68,11 +70,30 @@
...
 
-},{"./utils":2,"react":"CwoHg3"}],2:[function(_dereq_,module,exports){
-exports.extend = function (extandable, extending) {
-    for (var i in extending) {
-        if (!(i in extandable)) {
-            extandable[i] = extending[i];
+},{"./utils":3}],3:[function(_dereq_,module,exports){
+var getAllProperties = function (obj) {
+    var props = {};
+    while (obj && obj !== React.Component.prototype && obj !== Object.prototype) {
```
Comment (kof) [inline dist/reactive-elements.js]: Why is `React` not required, but used from global?

**Comment by timdp** [comment]
@kof I'm probably missing something here. I thought the idea of this code was to expose the React component's API through the custom element's.

As an example, if the hypothetical React component `Notification` has a function called `flash` that draws the user's attention, I expect the corresponding Reactive Element `<reactive-notification>` to let me write `document.querySelector('reactive-notification').flash()`. With `React.createClass`, that works fine.

However, in the case of ES6 classes, instance methods (as well as getters and setters) aren't enumerable, so they won't be picked up by the `for ... in`. How would you expose the `flash` method so that the developer does not need to access the Reactive Element's internals?

I completely agree that this is a very aggressive solution. I just needed something that worked. I'm certainly open to alternatives.

**Comment by kof** [comment]
Again ... you could explicitly expose your `flash` function before like this within react component: `this.props.element.flash = :this.flash`

Now you are doing a dangerous merge I got rid of some time ago ... I don't understand why @PixelsCommander merged this.

**Comment by kof** [comment]
It has been even documented https://github.com/PixelsCommander/ReactiveElements#exposing-components-methods-on-custom-element

**Comment by kof** [comment]
@PixelsCommander please remove this bs from code ...

**Comment by PixelsCommander** [comment]
This is true that exposed methods potentially confuse someone by overlapping native methods. However this is one of RE selling points and this is the reason @timdp picked it as a tool.

For me it looks cool to have auto binding and only problem I see here is that developer is not notified when it happens. Why not to check for overlapping and not notify by putting error into console instead of removing handy feature?

**Comment by kof** [comment]
I assume @timdp sent this PR because he didn't read documentation.

**Comment by PixelsCommander** [comment]
As I see it... this is not about documentation but about making handy and cool tool. I was involved into that stuff as well =)

**Comment by kof** [comment]
So I would not be able to have a .focus method within my component ? No this is not the right way.

**Comment by kof** [comment]
React component and DOM element are 2 different classes. We can't just melt them together without issues.

**Comment by kof** [comment]
We can only carefully provide an explicit way of extending DOM element.

**Comment by timdp** [comment]
@kof I indeed missed that link in the docs, so thanks. How about making it easier by providing a list of properties to proxy to the API?

There's no need for that tone of yours, by the way. I was just trying to contribute to a project that I appreciate. I get that you're upset, but you could've worded it differently.

**Comment by PixelsCommander** [comment]
Agree, but that is exactly what good tool does - removing issues and complexity. I propose autobind until it overlaps and fire notice then.

**Comment by timdp** [comment]
@PixelsCommander For what it's worth, I'm also in favor of a less aggressive API, as I said initially. Can we somehow provide an array of properties to proxy and run those through my code instead of proxying *all* properties by default?

**Comment by kof** [comment]
You still won't be able to have a .focus method just because your react component is passed as a custom element. 

A component should stay a component. It shouldn't be designed with custom element in mind.

**Comment by kof** [comment]
This is an a very bad implicit dependency.

**Comment by kof** [comment]
@timdp I am upset because every time I try to update ReactiveElements I find out that somebody broke it for me.

**Comment by timdp** [comment]
> You still won't be able to have a .focus method just because your react component is passed as a custom element.

Okay, so how would I get my hypothetical notification component to flash? A dummy property? An event on the window? I'm struggling to think of a good API for it.

**Comment by kof** [comment]
Component is Component, CustomElement is an Element. They are different. Evtl the method where we create  a binding should also pass a number of methods that need to be exposed from the component on the element. Component should not even know that he is wrapped by a custom element.

**Comment by kof** [comment]
@timdp read again https://github.com/PixelsCommander/ReactiveElements/pull/54#issuecomment-153712493

**Comment by timdp** [comment]
So in the end, you'd still add `flash` to the Reactive Element, but explicitly and only that one rather than copying *all* the properties, yes? I'm pretty sure we're in agreement.

**Comment by timdp** [comment]
Also, my original implementation didn't take getters and setters into account, so it'll need to be adapted anyway. I started patching it last week but I never finished it.

**Comment by kof** [comment]
> So in the end, you'd still add flash to the Reactive Element, but explicitly and only that one rather than copying all the properties, yes? I'm pretty sure we're in agreement.

exactly

>Also, my original implementation didn't take getters and setters into account, so it'll need to be adapted anyway. I started patching it last week but I never finished it.

thats why I would immediately remove all the auto merging magic and use an explicit solution only. Ideally a Component doesn't even know that it is wrapped into custom element.

**Comment by PixelsCommander** [comment]
Guys, lets make auto bind until it overlaps. And notify developer then. I do not see any disadvantages there, so it looks like a pretty healthy "merging magic" which will never break any code. Do you have objections on that?

**Comment by kof** [comment]
You can't notify a developer to stop using some method which conflicts with custom element. His component doesn't even know potentially that you wrapped his component into custom element.

**Comment by kof** [comment]
Those are 2 very different abstraction layers ...

**Comment by kof** [comment]
What you are trying to do is basically to make an Element inherit from a Component. This is a very very bad design. They are different by nature.

**Comment by kof** [comment]
Here is a much better way:

```
document.registerReact('my-react-component', MyComponent, {expose: ['myMethod1', 'myMethod2']});

```

**Comment by kof** [comment]
The code who uses a Component to register a custom element, defines methods to be exposed on custom elements. Here you can warn user if there are conflicts.

**Comment by timdp** [comment]
The options object to `registerReact` would also be my preferred approach.

That being said, nothing's stopping you from also having an `autobind` option that enables the current, flawed behavior.

**Comment by kof** [comment]
I would remove autobind completely, or at least make it disabled by default, `expose: 'all'` could turn on the magic if you absolutely can't leave without it.

**Comment by timdp** [comment]
I could certainly live without it. The only potential issue I can see is that the person *using* the component will need to declare what part of the wrapped component's API they will be using. However, I can't think of any other party who'd be responsible for that.

Maybe it would be useful if the author of a custom element could provide a "hint" to the Reactive Elements API, to suggest a list of properties to expose by default. I.e., a "Reactive Elements-ready component" could contain code along the lines of

```js
if (this.reactiveElement) {
  this.reactiveElement.init({ expose: [...] })
}
```

However, that would completely violate loose coupling, so it's not the most elegant approach. Just trying to think of ways to make people's lives easier though.

**Comment by kof** [comment]
There are only 2 parties envolved. Component and Application. Application developer can register any component as a custom element. Application developer KNOWS which methods he needs to use imperatively. There is no problem of explicit exposing  within .registerReact call.

**Comment by timdp** [comment]
That's true, I guess. I'm just worried that the `registerReact` call will get quite complicated.

Also, there will no doubt be properties that have different semantics in a component context than in an RE context. It might be better to let `expose` be an object than an array, which would let you pass transformers, e.g.:

```js
document.registerReact('reactive-notification', Notification, {
  expose: {
    // <reactive-notification> exposes .flash()
    flash: function () {  // Instance method proxy
      // Component exposes .startFlashing(time)
      return this.component.startFlashing(500 /* ms */)
    },

    // <reactive-notification> exposes .isFlashing getter and no setter
    isFlashing: {
      get: function () {  // Setter proxy
        return this.component.flashing
      },
      set: null  // Can be omitted
    },

    // Shorthand for this.component.doSomething.apply(this.component, arguments)
    doSomething: true
  }
})
```

Makes sense?

**Comment by kof** [comment]
This is a whole different problem you are trying to tackle. If you don't like how components method work, you need to fix or wrap component. 

Right now you are trying to mix methods exposing and methods overrides ... also some sort of OOP thing.

`return this.component.startFlashing(500 /* ms */)`

Very bad. You don't know what `this` is here.

If you need to change methods, you better wrap component with your own component and do there whatever you want and register YOUR component then as an reactive element.

**Comment by timdp** [comment]
It's a related problem though. You said it yourself: a component is not an element and vice versa, so their APIs have different semantics. If you want to translate one to the other, you're going to end up writing some conversion logic, and Reactive Elements might as well take care of that, since it's the very library that facilitates the translation.

About the `this`: it might as well be the first argument to the callback. That's hardly the point.

**Comment by kof** [comment]
> you're going to end up writing some conversion logic

yes I suggested to wrap the component into another component ... instead of creating a special API for for overrides.

**Comment by kof** [comment]
> About the this: it might as well be the first argument to the callback. That's hardly the point.

could be an optional way of exposing methods, which are not related to a component. If component is just an argument -> thats ok.

**Comment by timdp** [comment]
Wrapping a component in another component doesn't make sense to me, because that would mean that you're building a component with the semantics of a Custom Element. I still think the appropriate place for such translation logic is the glue code that Reactive Elements delivers. But I'd be happy to write a separate library that basically overrides `registerReact` with an enhanced version, to make it an optional feature.

**Comment by kof** [comment]
Oh I didn't mean it to be a translation logic. I ment it this way: If you are using a component and you want to change something in the way how it works - you can wrap this component into a different component. This would also work for the case when you need to use this component as a reactive element. The wrapper component would be  a special component for reactiveElement.

We are talking about different hypothetical situations. Also normally its a bad way of thinking .... a real use case is what we need there.

My component wrapper solution requires no changes in reactive-elements. Thats why its good as long as there is no real life use case for this functions map.

**Comment by kof** [comment]
There is one more problem with a suggested API. There is such a thing called "higher order component". A higher order component may hide public API of the inner component. In this case, only lower order component can expose its methods on custom element.

So the way I am using right now should remain for such cases.

**Pull request merged**